include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -0,0 +1,317 @@
+#include "sparrow/utils/contracts.hpp"
+#include "sparrow/utils/mp_utils.hpp"
+#include "sparrow/utils/variant_visitor.hpp"
+
+namespace sparrow
+{
+    class arrow_proxy
+    {
+    public:
+
+        explicit arrow_proxy(ArrowArray&& array, ArrowSchema&& schema);
+        explicit arrow_proxy(ArrowArray&& array, ArrowSchema* schema);
+        explicit arrow_proxy(ArrowArray* array, ArrowSchema* schema);
+
+        ~arrow_proxy();
+
+        [[nodiscard]] std::string_view format() const;
+        [[nodiscard]] std::string_view name() const;
+        [[nodiscard]] std::string_view metadata() const;
+        [[nodiscard]] std::vector<ArrowFlag> flags() const;
+        [[nodiscard]] int64_t length() const;
+        [[nodiscard]] int64_t null_count() const;
+        [[nodiscard]] int64_t offset() const;
+        [[nodiscard]] int64_t n_buffers() const;
+        [[nodiscard]] int64_t n_children() const;
+        [[nodiscard]] const std::vector<sparrow::buffer_view<uint8_t>>& buffers() const;
+        [[nodiscard]] const std::vector<arrow_proxy>& children() const;
+        [[nodiscard]] std::optional<arrow_proxy> dictionary() const;
+
+        [[nodiscard]] bool is_created_with_sparrow() const;
+
+        [[nodiscard]] void* private_data() const;
+
+    private:
+
+        std::variant<ArrowArray*, ArrowArray> m_array;
+        std::variant<ArrowSchema*, ArrowSchema> m_schema;
+        std::vector<sparrow::buffer_view<uint8_t>> m_buffers;
+        std::vector<arrow_proxy> m_children;
+
+        void initialize_children();
+        void initialize_buffers();
+
+        void validate_array_and_schema() const;
+
+        [[nodiscard]] ArrowArray& array();
+        [[nodiscard]] const ArrowArray& array() const;
+
+        [[nodiscard]] ArrowSchema& schema();
+        [[nodiscard]] const ArrowSchema& schema() const;
+
+        [[nodiscard]] std::size_t
+        compute_buffer_size(buffer_type buffer_type, int64_t length, data_type data_type) const;
+    };
+
+    void arrow_proxy::initialize_children()
+    {
+        m_children.reserve(static_cast<std::size_t>(array().n_children));
+        for (int64_t i = 0; i < array().n_children; ++i)
+        {
+            m_children.emplace_back(array().children[i], schema().children[i]);
+        }
+    }
+
+    void arrow_proxy::initialize_buffers()
+    {
+        m_buffers.clear();
+        const auto buffer_count = static_cast<size_t>(array().n_buffers);
+        m_buffers.reserve(buffer_count);
+        const data_type data_type = format_to_data_type(schema().format);
+        const std::vector<buffer_type> buffers_type = get_buffer_types_from_data_type(data_type);
+        for (std::size_t i = 0; i < buffer_count; ++i)
+        {
+            const auto buffer_type = buffers_type[i];
+            auto buffer = array().buffers[i];
+            const std::size_t buffer_size = compute_buffer_size(buffer_type, length(), data_type);
+            auto* ptr = reinterpret_cast<uint8_t*>(const_cast<void*>(buffer));
```

Comment:
Is the `reinterpret_cast` necessary? arent they supposed to already be `uint8_t` arrays?

Reply:
Fixed